flink-1.17-connector-bigquery/flink-connector-bigquery/src/main/java/com/google/cloud/flink/bigquery/sink/serializer/BigQuerySchemaProviderTableImpl.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2024 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.flink.bigquery.sink.serializer;
+
+import org.apache.flink.table.api.TableDescriptor;
+import org.apache.flink.table.types.DataType;
+
+import com.google.api.client.util.Preconditions;
+import com.google.api.services.bigquery.model.TableSchema;
+import com.google.cloud.flink.bigquery.common.config.BigQueryConnectOptions;
+import com.google.cloud.flink.bigquery.table.config.BigQueryConnectorOptions;
+import com.google.protobuf.DescriptorProtos;
+import com.google.protobuf.Descriptors;
+import org.apache.avro.Schema;
+
+import java.io.IOException;
+import java.util.Map;
+
+/**
+ * Default implementation of {@link BigQuerySchemaProvider} deriving Avro {@link Schema} from {@link
+ * TableSchema}, which in turn is sourced from {@link BigQueryConnectOptions}.
+ */
+public class BigQuerySchemaProviderTableImpl implements BigQuerySchemaProvider {
+
+    private final TableDescriptor tableDescriptor;
+    private BigQuerySchemaProvider bigQuerySchemaProvider;
+
+    private BigQueryConnectOptions getConnectOptionsFromMap(Map<String, String> options)
+            throws IOException {
+        return BigQueryConnectOptions.builder()
+                .setTable(options.get(BigQueryConnectorOptions.TABLE.key()))
+                .setProjectId(options.get(BigQueryConnectorOptions.PROJECT.key()))
+                .setDataset(options.get(BigQueryConnectorOptions.DATASET.key()))
+                .build();
+    }
+
+    public static DataType getDataTypeSchemaFromAvroSchema(Schema avroSchema) {
+        return AvroSchemaConvertor.convertToDataType(avroSchema.toString());
+    }
+
+    private org.apache.flink.table.api.Schema getTableApiSchemaFromAvroSchema(Schema avroSchema) {
+        DataType dataTypeSchema = getDataTypeSchemaFromAvroSchema(avroSchema);
+        return org.apache.flink.table.api.Schema.newBuilder()
+                .fromRowDataType(dataTypeSchema)
+                .build();
+    }
+
+    private org.apache.flink.table.api.Schema getTableApiSchemaFromAvroSchema() {
+        Schema avroSchema = this.getAvroSchema();
+        if (avroSchema == null) {
+            throw new RuntimeException(
+                    "Avro Schema Not initialized before obtaining Table API schema");
+        }
```

Comment:
you already have avroSchema variable defined at this point. can use that instead of calling this.getAvroSchema() again

Reply:
Fixed.